nengo/tests/test_processes.py

```diff
@@ -334,3 +334,185 @@ def test_present_input(Simulator, rng):
     for k, [ii, image] in enumerate(zip(i, y)):
         assert np.allclose(image, images[ii], rtol=1e-4, atol=1e-7), (k, ii)
+
+
+class TestPiecewise(object):
+
+    dt = 0.001
+
+    def run_sim(self, data, interpolation, Simulator):
+        process = nengo.processes.Piecewise(data, interpolation)
+
+        with nengo.Network() as model:
+            u = nengo.Node(process, size_out=process.default_size_out)
+            up = nengo.Probe(u)
+
+        with Simulator(model, dt=self.dt) as sim:
+            sim.run(1.5)
+
+        return sim.data[up]
+
+    def test_basic(self, Simulator):
+        f = self.run_sim({0.5: 1, 1.0: 0}, 'zero', Simulator)
+        assert np.allclose(f[0], [0])   # t = 0.001
+        assert np.allclose(f[249], [0]) # t = 0.25
+        assert np.allclose(f[498], [0]) # t = 0.499
+        assert np.allclose(f[499], [1]) # t = 0.5
+        assert np.allclose(f[749], [1]) # t = 0.75
+        assert np.allclose(f[999], [0]) # t = 1.0
+        assert np.allclose(f[1499], [0]) # t = 1.5
+
+
+    def test_lists(self, Simulator):
+        f = self.run_sim({0.5: [1, 0], 1.0: [0, 1]}, 'zero', Simulator)
+        assert np.allclose(f[0], [0, 0])   # t = 0.001
+        assert np.allclose(f[249], [0, 0]) # t = 0.25
+        assert np.allclose(f[498], [0, 0]) # t = 0.499
+        assert np.allclose(f[499], [1, 0]) # t = 0.5
+        assert np.allclose(f[749], [1, 0]) # t = 0.75
+        assert np.allclose(f[999], [0, 1]) # t = 1.0
+        assert np.allclose(f[1499], [0, 1]) # t = 1.5
+
+
+
+    def get_step(self, data):
+        process = nengo.processes.Piecewise(data)
+        return process.make_step(shape_in=(process.default_size_in,),
+                                 shape_out=(process.default_size_out,),
+                                 dt=0.001, rng=None)
+
+
+    def test_default_zero(self):
+        f = self.get_step({0.5: 1, 1.0: 0})
+        assert np.allclose(f(-10), [0])
+        assert np.allclose(f(0), [0])
+
+        f = self.get_step({0.5: [1, 0], 1.0: [0, 1]})
+        assert np.allclose(f(-10), [0, 0])
+        assert np.allclose(f(0), [0, 0 ])
+
+
+    def test_invalid_key(self):
+        data = {0.5: 1, 1: 0, 'a': 0.2}
+        with pytest.raises(ValidationError):
+            process = nengo.processes.Piecewise(data)
+            assert process
+
+
+    def test_invalid_length(self):
+        data = {0.5: [1, 0], 1.0: [1, 0, 0]}
+        with pytest.raises(ValidationError):
+            process = nengo.processes.Piecewise(data)
+            assert process
+
+
+    def test_invalid_function_length(self):
+        data = {0.5: 0, 1.0: lambda t: [t, t ** 2]}
+        tp, yp = zip(*data.items())
+        with pytest.raises(ValidationError):
+            process = nengo.processes.Piecewise(tp, yp)
+            assert process
+
+
+    def test_function(self, Simulator):
+        f = self.run_sim({0: np.sin, 0.5: np.cos}, 'zero', Simulator)
+
+        assert np.allclose(f[0], [np.sin(0.001)])   # t = 0.001
```

Comment:
As I did in the other other tests, we should opt for passing in things like `t == 0.001` rather than passing in `0` and commenting the actual time (which is more understandable to the reader).